File: javascript/euler39.js
/* 

  Euler #39 in JavaScript.

  """
  If p is the perimeter of a right angle triangle with integral length sides, 
  {a,b,c}, there are exactly three solutions for p = 120.
   
  {20,48,52}, {24,45,51}, {30,40,50}
   
  For which value of p <= 1000, is the number of solutions maximised?
  """


  This JavaScript model was created by Hakan Kjellerstrand, hakank@gmail.com
  See also my JavaScript page: http://www.hakank.org/javascript/

*/

'use strict';
const {range2,sum2,max2,timing2} = require('./js_utils.js');

// 22ms
var euler39a = function() {
    var n = 1000;
    var squares = new Set(range2(1,n).map(i=>i*i));
    var valid = [];
    for (var x of squares) {
        for (var y of squares) {
            if (x < y && squares.has(x+y) && (Math.sqrt(x) + Math.sqrt(x) + Math.sqrt(x+y)) < 1000) {
                valid.push([x,y]);
            }
        }
    }
    var counts = {};
    for(var [x2,y2] of valid) {
       var c = Math.floor(Math.sqrt(x2) + Math.sqrt(y2) + Math.sqrt(x2+y2));
       if (!counts[c]) {
           counts[c] = 0;
       }
       counts[c] += 1;
    }

    // find max count
    var maxV = Object.values(counts).map(i=>parseInt(i)).max2();
    return Object.keys(counts)
    .filter(i=>counts[i] === maxV);

 
}

timing2(euler39a); // 22ms

File: julia/euler39.jl
#=

  Euler #39 in Julia.

  """
  If p is the perimeter of a right angle triangle with integral length sides,
  {a,b,c}, there are exactly three solutions for p = 120.

  {20,48,52}, {24,45,51}, {30,40,50}

  For which value of p <= 1000, is the number of solutions maximised?
  """

  This Julia model was created by Hakan Kjellerstrand, hakank@gmail.com
  See also my Julia page: http://www.hakank.org/julia/

=#

include("Euler.jl")

# 0.11346191s
# This is surprinsingly slow!
function euler39a()
    n = 1000;
    # squares = Set((1:n).|>i->i*i) # slightly slower
    squares = Dict((1:n).|>i->i*i=>1)
    valid = []
    for x in keys(squares)
        for y in keys(squares)
            if x < y && haskey(squares,x+y) && (sqrt(x) + sqrt(x) + sqrt(x+y)) < 1000
                push!(valid,[x, y])
            end
        end
    end

    counts = Dict()
    for x in valid
        c = floor(Int,sqrt(x[1]) + sqrt(x[2]) + sqrt(x[1]+x[2]))
        get!(counts,c,0)
        counts[c] += 1
   end

    # find max count
    maxV = maximum(values(counts))
    t = [i for i in keys(counts) if counts[i] == maxV]
    return t[1]
end

run_euler(euler39a)
